Functional Alternative Splicing Events Score (FAScore)
================
</br>

## What is FAScore?

FAScore is a machine learning framework designed to predict functionally significant alternative splicing (AS) events during biological processes such as hematopoietic differentiation. By integrating dynamic expression patterns of AS events and their host genes with isoform-specific sequence conservation and structural features, it computes functional scores through Random Forest (RF) modeling. The framework further classifies AS events into functional categories using a Gaussian Mixture Model (GMM). The analysis incorporates two feature classes:

Dynamic features: Expression breadth (*R*), tissue specificity (*τ*), host gene correlation (Spearman’s *ρ* with significance *P*<sub>*ρ*</sub>), and temporal expression trends (regression slope *β* with *P*<sub>*β*</sub>). These are normalized (0-1 scale) and combined into a DyScore (-1 to 1), where positive/negative values indicate up-/down-regulation respectively. Structural features: Derived from APPRIS, including evolutionary conservation, functional domains, transmembrane helices, and subcellular localization signatures.

FAScore outputs both DyScore (dynamic score) and FAScore (functional score) for all annotated AS events. The package supports full analysis in 11 species (human, macaque, chimpanzee, mouse, rat, pig, cow, chicken, zebrafish, fruitfly and C. elegans) and DyScore calculation in any species.

</br>

<div align="center">
<img width = '900' height ='320' src ="https://github.com/LuChenLab/FAScore/blob/master/figure/workflow.png"/>
</div>

<br>



## Contents

[Installation](#Installation)  
[Input](#Input)  
[Output](#Output)  
[Tutorial](#Tutorial)  
[SessionInfo](#SessionInfo)  
[Contact](#Contact)  
[Citation](#Citation)

## Installation

FAScore has been developed with `R 4.0.0` and the following packages are needed to be installed in `R` (R scripts in FAScore automatically check to see if an R library is already installed and then install those that are needed. So no need for manual preinstallation!):

    randomForest (>= 4.0),
    SummarizedExperiment (>= 1.18),
    dplyr (>= 1.1),
    GenomicRanges (>= 1.43),
    GenomicFeatures (>= 1.43),
    ggplot2 (>= 3.5),
    ggrepel (>= 0.9),
    methods (>= 4.0),
    rtracklayer (>= 1.51),
    stringr (>= 1.5),
    tidyr (>= 1.3),
    IRanges (>= 2.25),
    stats (>= 4.0),
    reshape2 (>= 1.0),
    parallel (>= 4.0),
    S4Vectors (>= 0.26),
    cowplot (>= 1.1),
    mclust (>= 5.4)

To install FAScore, you have two options: either install directly from GitHub or use the compressed source file:

``` r
# Install from GitHub if remotes package is not installed
if (!requireNamespace("remotes", quietly = TRUE))
    install.packages("remotes")

remotes::install_github("LuChenLab/FAScore/package")
```

Alternatively, you can install FAScore using the source file downloaded from the repository :

``` bash
# Install FAScore from a downloaded source file
R CMD INSTALL FAScore_0.3.0.tar.gz
```

</br> </br>

## Input

#### PSI value of all AS events

A data frame where rows represent alternative splicing (AS) events or splice junctions (SJs), columns correspond to sample identifiers, and values are pre-computed percent spliced in (PSI) scores generated by tools such as rMATS or ICAS.

#### Gene expression

A data frame containing normalized gene expression counts (e.g., TPM, FPKM, or DESeq2-normalized counts) with gene identifiers as rows and matching sample identifiers (consistent with the PSI data frame) as columns.

#### Isoform expression

A data frame containing normalized transcript-level expression counts (e.g., RSEM-derived TPM/FPKM values), where rows represent transcript identifiers, columns correspond to sample identifiers (must match those in the PSI data frame), and values contain the normalized isoform expression measurements.

#### Gene annotation file (GTF file)

A file path pointing to a Gene Transfer Format (GTF) file containing genome annotations for alternative splicing (AS) events, splice junctions (SJs), isoforms, and gene loci. Species-specific GTF files are available from the [**Ensembl**](https://asia.ensembl.org/info/data/ftp/index.html) database

#### Sample information

A data frame where rows represent sample identifiers (matching column names in PSI, gene expression, and isoform expression matrices), containing a required grouping variable with at least three distinct biological classes, along with any additional sample metadata.

#### Isoform structural information (optional)

A data frame with structural-related scores such as cross-species sequence conservation score (Conservation), the functionally important residues score (AA.residues), domain integrity score (Domain.integrity), the structural homologs and integrity score (Protein3D), trans-membrane helices score (Transmembrane), signal peptide score (SignalP), and subcellular location score (TargetP). Files for species beyond **human**(GRCh38), **rhesus macaque**(Mmul10), **rat**(Rnor6.0), **mouse**(GRCm38), and **zebrafish**(GRCz11) can be downloaded from the[**APPRIS**](https://appris.bioinfo.cnio.es/#/) database. </br> </br>

## Output

The output will be an S4 object containing the following slots:

-   `assays` : A SimpleList of matrix-like objects, including the PSI matrix of AS events derived from the input.
-   `colData` : An optional DataFrame providing information on the samples, as specified in the input.
-   `rowRanges` : A GRanges object describing the ranges of AS events and corresponding gene and isoform IDs. It includes columns for **ASID**, **TranscriptID**, **GeneID**, **GeneName** and **matchTransID**. The **matchTransID** column stores the best-matching transcript for each gene based on maximum expression values (from `ISOFORM` data), aiding in mapping isoform structural information.
-   `GENE` : A data frame of gene expression values provided in the input.
-   `ISOFORM` : A data frame of isoform expression values provided in the input.
-   `Correlation` : A list containing the Spearman’s rank correlation coefficient (*ρ*) and corresponding p-value (*P*<sub>*ρ*</sub>) for genes or AS events.
-   `Tau` : A list containing the specificity index (*τ*) for genes and AS events, which quantifies the specificity of expression profiles. Higher values indicate greater specificity.
-   `Range` : A list containing range value (*R*) for genes and AS events, representing the breadth of expression.
-   `Linear` : A list containing linear fit parameters (intercept, slope (*β*), r<sup>2</sup>), and the p-value (*P*<sub>*β*</sub>) for the linear fit of genes and AS events, representing the expression gradient.
-   `DyScore` : A list capturing dynamic scores for genes and AS events.
-   `Structure` : A data frame of isoform structural scores obtained from the APPRIS database.
-   `RFpredict` : The predicted scores and classification results from the pre-trained model, stored in the **FAScore** and **FAStype** columns.
-   `GMM` : The parameter result of gaussian mixture model (GMM).

</br> </br>

## Tutorial

### Usage

For detailed usage of parameters of functions, please `help()` or `?`.

### Prepare

Loading example dataset (You can download from our **figshare** database: <https://doi.org/10.6084/m9.figshare.29069336.v2>).

``` r
library(FAScore)
library(dplyr)
library(ggplot2)
```

``` r
df_AS <- read.table("MusMus_Erythroid_AS.txt", header = T, check.names = FALSE)
df_AS <- df_AS[apply(df_AS, 1, function(x) {sum(x >= 0.05 & x <= 0.95, na.rm = T) >= 2}), ]
df_Gene <- read.table("MusMus_Erythroid_Gene.txt", header = T, check.names = FALSE)
df_Iso <- read.table("MusMus_Erythroid_Iso.txt", header = T, check.names = FALSE)
meta <- readRDS("metainfo.Rds")
```

``` r
head(df_AS)
```

    ##                         LT-HSC_1 LT-HSC_2  ST-HSC_1  ST-HSC_2
    ## X:161255030-161255659:+ 0.396732 0.365764 0.2242840 0.4305620
    ## 6:17282083-17282684:+   0.135535 0.180492 0.0843074 0.0553682
    ## 4:120456285-120456509:+ 0.113175 0.375139 0.2064740 0.3071080
    ## 4:120462013-120462079:+ 0.602268 0.734503 0.5848140 0.6001620
    ## 4:120462080-120462155:+ 0.616394 0.761759 0.6365940 0.6578870
    ## 4:120463075-120463084:+ 0.922318 0.961764 0.9647620 0.9496600
    ##                             MPP_1    MPP_2    CMP_1    CMP_2
    ## X:161255030-161255659:+ 0.4158790 0.477915 0.409742 0.439201
    ## 6:17282083-17282684:+   0.0630537 0.128031 0.168954 0.440274
    ## 4:120456285-120456509:+ 0.1488300 0.386593 0.143237 0.373457
    ## 4:120462013-120462079:+ 0.5240050 0.643911 0.554912 0.652065
    ## 4:120462080-120462155:+ 0.5595110 0.704905 0.650354 0.662815
    ## 4:120463075-120463084:+ 0.9711360 0.962019 0.946724 0.944725
    ##                            MEP_1    MEP_2     MK_1     MK_2
    ## X:161255030-161255659:+ 0.397258 0.378395 0.215838 0.479469
    ## 6:17282083-17282684:+   0.000000 0.000000 0.103275 0.101225
    ## 4:120456285-120456509:+ 0.205100 0.254499 0.218328 0.482626
    ## 4:120462013-120462079:+ 0.607083 0.636477 0.687440 0.800307
    ## 4:120462080-120462155:+ 0.621688 0.635275 0.663026 0.814104
    ## 4:120463075-120463084:+ 0.947275 0.923054 0.954565 0.974340
    ##                           EryA_1    EryA_2   EryB_1   EryB_2
    ## X:161255030-161255659:+ 0.424259 0.5186920 0.390533 0.657807
    ## 6:17282083-17282684:+   0.124026 0.0374126 0.000000 0.000000
    ## 4:120456285-120456509:+ 0.348533 0.2273770 0.278966 0.137418
    ## 4:120462013-120462079:+ 0.665422 0.6070830 0.661961 0.785945
    ## 4:120462080-120462155:+ 0.721278 0.6774960 0.741367 0.820738
    ## 4:120463075-120463084:+ 0.970910 0.9372700 1.000000 1.000000

``` r
head(df_Gene)
```

    ##                    LT-HSC_1 LT-HSC_2 ST-HSC_1 ST-HSC_2 MPP_1
    ## ENSMUSG00000000001    42.81    37.39    42.40    30.81 28.19
    ## ENSMUSG00000000003     0.00     0.00     0.00     0.00  0.00
    ## ENSMUSG00000000028    14.89     6.48    15.21     9.79  9.74
    ## ENSMUSG00000000031     0.88     0.45     0.09     0.40  0.02
    ## ENSMUSG00000000037     1.06     1.87     1.33     2.79  1.16
    ## ENSMUSG00000000049     0.00     0.00     0.00     0.00  0.00
    ##                    MPP_2 CMP_1 CMP_2 MEP_1 MEP_2  MK_1  MK_2
    ## ENSMUSG00000000001 26.35 31.50 18.98 23.66 14.39 28.02 31.39
    ## ENSMUSG00000000003  0.00  0.00  0.00  0.00  0.00  0.00  0.00
    ## ENSMUSG00000000028  9.74 10.19  8.01 13.28  9.99  6.65  6.33
    ## ENSMUSG00000000031  0.99  0.11  0.10  0.03  0.00  0.01  0.02
    ## ENSMUSG00000000037  1.81  0.93  0.93  2.03  2.00  0.21  0.48
    ## ENSMUSG00000000049  0.00  0.00  0.00  0.00  0.00  0.00  0.02
    ##                    EryA_1 EryA_2 EryB_1 EryB_2
    ## ENSMUSG00000000001  15.55  19.40   3.62   3.29
    ## ENSMUSG00000000003   0.00   0.00   0.00   0.00
    ## ENSMUSG00000000028  14.69  20.53   6.00   4.83
    ## ENSMUSG00000000031   0.01   0.01   0.00   0.00
    ## ENSMUSG00000000037   0.77   0.91   0.14   0.10
    ## ENSMUSG00000000049   0.00   0.00   0.00   0.00

``` r
head(df_Iso)
```

    ##                               gene_id LT-HSC_1 LT-HSC_2 ST-HSC_1
    ## ENSMUST00000000001 ENSMUSG00000000001    42.81    37.39    42.40
    ## ENSMUST00000000003 ENSMUSG00000000003     0.00     0.00     0.00
    ## ENSMUST00000000010 ENSMUSG00000020875     0.07     0.06     0.00
    ## ENSMUST00000000028 ENSMUSG00000000028    13.02     5.13    13.30
    ## ENSMUST00000000033 ENSMUSG00000048583     0.30     0.12     0.01
    ## ENSMUST00000000049 ENSMUSG00000000049     0.00     0.00     0.00
    ##                    ST-HSC_2 MPP_1 MPP_2 CMP_1 CMP_2 MEP_1 MEP_2
    ## ENSMUST00000000001    30.81 28.19 26.35 31.50 18.98 23.66 14.39
    ## ENSMUST00000000003     0.00  0.00  0.00  0.00  0.00  0.00  0.00
    ## ENSMUST00000000010     0.00  0.00  0.00  0.00  0.00  0.00  0.00
    ## ENSMUST00000000028     8.03  8.63  8.49  9.04  6.51 11.44  7.85
    ## ENSMUST00000000033     0.05  0.00  0.00  0.03  0.01  0.00  0.00
    ## ENSMUST00000000049     0.00  0.00  0.00  0.00  0.00  0.00  0.00
    ##                     MK_1  MK_2 EryA_1 EryA_2 EryB_1 EryB_2
    ## ENSMUST00000000001 28.02 31.39  15.55  19.40   3.62   3.29
    ## ENSMUST00000000003  0.00  0.00   0.00   0.00   0.00   0.00
    ## ENSMUST00000000010  0.00  0.00   0.00   0.00   0.00   0.00
    ## ENSMUST00000000028  4.78  4.20  11.57  16.06   4.26   3.44
    ## ENSMUST00000000033  0.00  0.02   0.00   0.00   0.00   0.00
    ## ENSMUST00000000049  0.00  0.02   0.00   0.00   0.00   0.00

``` r
data(ExDataSet)
head(ExDataSet$meta)
```

    ##               Run CellType    Lineage
    ## LT-HSC_1 LT-HSC_1   LT-HSC       Stem
    ## LT-HSC_2 LT-HSC_2   LT-HSC       Stem
    ## ST-HSC_1 ST-HSC_1   ST-HSC       Stem
    ## ST-HSC_2 ST-HSC_2   ST-HSC       Stem
    ## MPP_1       MPP_1      MPP Progenitor
    ## MPP_2       MPP_2      MPP Progenitor

</br>

### Creating a FAScore object

The alternative splicing PSI matrix, gene expression matrix, transcripts expression matrix, and sample information table need to be prepared to create a FAScore object.

``` r
MyObj <- FAScoreDataSet(colData = ExDataSet$meta, AS = df_AS, 
                        GENE = df_Gene, ISOFORM = df_Iso)
```

``` r
MyObj
```

    ## class: FAScore 
    ## dim: 22885 16 
    ## metadata(1): version
    ## assays(1): AS
    ## rownames(22885): X:161255030-161255659:+
    ##   6:17282083-17282684:+ ... 16:92076720-92087473:+
    ##   16:91676808-91676923:-
    ## rowData names(0):
    ## colnames(16): LT-HSC_1 LT-HSC_2 ... EryB_1 EryB_2
    ## colData names(3): Run CellType Lineage

</br>

### Calculate the dynamic scores of AS events and genes

The dynamic score of genes or AS events is computed as the mean of scaled (0-1) dynamic features from `CalcuFeature`, including: R<sub>scaled</sub> (normalized TPM/PSI range), *τ* (specificity index, 0-1), *ρ* (Spearman’s correlation) with binarized *P*<sub>*ρ*</sub>, and *β* (linear slope, -1 to 1) with binarized *P*<sub>*β*</sub>. The final score ranges from -1 to 1, where absolute values represent effect size (0-1) and signs indicate directionality (*β* slope). This calculation is performed by the `CalcuDyScore` function.

 </br>

Preprocessing the related dynamic features

``` r
MyObj <- CalcuFeature(MyObj, group.by = "CellType", cores = 10)
```

``` r
head(MyObj@Correlation$Gene)  # head(MyObj@Correlation$AS)
```

    ##                    Spearman.cor   Spearman.p
    ## ENSMUSG00000000001   -0.7929133 0.0002483022
    ## ENSMUSG00000000003           NA           NA
    ## ENSMUSG00000000028   -0.2842377 0.2860039065
    ## ENSMUSG00000000031   -0.8488461 0.0000321010
    ## ENSMUSG00000000037   -0.7224375 0.0015723392
    ## ENSMUSG00000000049    0.1690309 0.5314416160

``` r
head(MyObj@Tau$Gene)  # head(MyObj@Tau$AS)
```

    ## ENSMUSG00000000001 ENSMUSG00000000003 ENSMUSG00000000028 
    ##          0.4343605          0.0000000          0.4681188 
    ## ENSMUSG00000000031 ENSMUSG00000000037 ENSMUSG00000000049 
    ##          0.8077336          0.5006935          1.0000000

``` r
head(MyObj@Range$Gene)  # head(MyObj@Range$AS)
```

    ## ENSMUSG00000000001 ENSMUSG00000000003 ENSMUSG00000000028 
    ##             36.645              0.000             12.195 
    ## ENSMUSG00000000031 ENSMUSG00000000037 ENSMUSG00000000049 
    ##              0.665              1.940              0.010

``` r
head(MyObj@Linear$Gene)  # head(MyObj@Linear$AS)
```

    ##                     intercept        slope         r2      pvalue
    ## ENSMUSG00000000001 29.2785221 -0.535831428 0.44054084 0.005054309
    ## ENSMUSG00000000003  0.0000000  0.000000000        NaN         NaN
    ## ENSMUSG00000000028 19.3882798 -0.068948676 0.02621270 0.549136371
    ## ENSMUSG00000000031  2.8080900 -0.179342917 0.45241263 0.004304452
    ## ENSMUSG00000000037  7.8488475 -0.250208237 0.33808048 0.018155151
    ## ENSMUSG00000000049 -0.0023659  0.001351943 0.05517241 0.381202376

</br>

Calculating the dynamic scores

``` r
MyObj <- CalcuDyScore(MyObj, maxRange = 10, maxSlope = 1, type = "Gene")
MyObj <- CalcuDyScore(MyObj, maxRange = 1, maxSlope = 1, type = "AS")
```

``` r
head(MyObj@DyScore$Gene)
```

    ## ENSMUSG00000000001 ENSMUSG00000000003 ENSMUSG00000000028 
    ##         -0.7938509                 NA         -0.3035509 
    ## ENSMUSG00000000031 ENSMUSG00000000037 ENSMUSG00000000049 
    ##         -0.6393204         -0.5788899          0.1950638

``` r
head(MyObj@DyScore$AS)
```

    ## X:161255030-161255659:+   6:17282083-17282684:+ 
    ##              0.09694379             -0.41579874 
    ## 4:120456285-120456509:+ 4:120462013-120462079:+ 
    ##              0.04968903              0.10764239 
    ## 4:120462080-120462155:+ 4:120463075-120463084:+ 
    ##              0.27491026              0.07131412

</br>

### Calculate the functional scores of AS events

The framework integrates dynamic, evolutionary conserved, and structural features to train a predictive model using randomForest (provided as a
pre-trained model). Functional scores and classifications are generated via the `CalcuFAScore` function, which supports both the pre-trained model and user-customized models.

</br>

Adding gene and transcript information of AS events. Note: Processing time may increase significantly with larger FAScore object dimensions.

``` r
my_gtf <- "Mus_musculus.GRCm38.101.gtf"
# or using other gtf files


MyObj <- ASmapIso(MyObj, gtf = my_gtf, AStype = "exonic", cores = 10)
MyObj <- ChooseIso(MyObj)
```

</br>

Adding the structural scores of transcripts

``` r
MyObj <- matchAppris(MyObj, gtf = my_gtf, species = "MusMus")

head(MyObj@Structure)
```

    ##               GeneID       TranscriptID                    ASID
    ## 1 ENSMUSG00000000085 ENSMUST00000000087 4:120462080-120462155:+
    ## 2 ENSMUSG00000000085 ENSMUST00000000087 4:120529260-120529753:+
    ## 3 ENSMUSG00000000085 ENSMUST00000000087 4:120462013-120462079:+
    ## 4 ENSMUSG00000000085 ENSMUST00000000087 4:120529249-120529259:+
    ## 5 ENSMUSG00000000085 ENSMUST00000000087 4:120463075-120463084:+
    ## 6 ENSMUSG00000000085 ENSMUST00000000087 4:120463085-120463145:+
    ##   Firestar Matador3D CORSAIR SPADE THUMP SignalP TargetP
    ## 1        0      7.46     4.8 390.8     0      -3      -6
    ## 2        0      7.46     4.8 390.8     0      -3      -6
    ## 3        0      7.46     4.8 390.8     0      -3      -6
    ## 4        0      7.46     4.8 390.8     0      -3      -6
    ## 5        0      7.46     4.8 390.8     0      -3      -6
    ## 6        0      7.46     4.8 390.8     0      -3      -6

</br>

or other species and annotation versions downloaded from [**APPRIS**](https://appris.bioinfo.cnio.es/#/) database

``` r
MyObj <- matchAppris(MyObj, gtf = my_gtf, appris = OtherFile)

head(MyObj@Structure)
```

</br>

Calculating the functional scores and classes

``` r
premodel = readRDS("finalModel.Rds")
MyObj <- CalcuFAScore(MyObj, model = premodel)

# Sorted by FAScore
MyObj@RFpredict <- MyObj@RFpredict[order(MyObj@RFpredict$FAScore,decreasing = T),]
```

</br>

Classify the AS by GMM

``` r
MyObj <- calcuGMM(MyObj, LOG = 2)
```

    ## fitting ...
    ##   |                                                   |   0%  |                                                                |============================                            |  50%  |                                                                |========================================================| 100%

``` r
head(MyObj@RFpredict)
```

    ##                           ASID
    ## 16712    8:95450737-95453349:-
    ## 18830   11:70440675-70440755:+
    ## 17765  3:126841485-126841573:+
    ## 4751    15:64312476-64312639:-
    ## 6137      19:7341460-7341498:-
    ## 4258  13:106987953-106988009:-
    ##                                                                                                                                                  TranscriptID
    ## 16712                                                                                                                   ENSMUST00000212214,ENSMUST00000212750
    ## 18830                                                                                                                                      ENSMUST00000102564
    ## 17765 ENSMUST00000199300,ENSMUST00000145454,ENSMUST00000131869,ENSMUST00000200171,ENSMUST00000171289,ENSMUST00000129293,ENSMUST00000167985,ENSMUST00000167417
    ## 4751                                                                                                 ENSMUST00000177371,ENSMUST00000177083,ENSMUST00000175710
    ## 6137                                                                              ENSMUST00000032557,ENSMUST00000164205,ENSMUST00000051711,ENSMUST00000165286
    ## 4258                                                                              ENSMUST00000022204,ENSMUST00000122233,ENSMUST00000159772,ENSMUST00000117539
    ##                                                                                TranscriptName
    ## 16712                                                                 Csnk2a2-203,Csnk2a2-209
    ## 18830                                                                               Arrb2-204
    ## 17765 Camk2d-226,Camk2d-213,Camk2d-209,Camk2d-228,Camk2d-221,Camk2d-207,Camk2d-218,Camk2d-217
    ## 4751                                                            Asap1-217,Asap1-216,Asap1-204
    ## 6137                                                  Mark2-202,Mark2-206,Mark2-203,Mark2-207
    ## 4258                                                  Kif2a-201,Kif2a-204,Kif2a-206,Kif2a-203
    ##                   GeneID GeneName       matchTransID
    ## 16712 ENSMUSG00000046707  Csnk2a2 ENSMUST00000212214
    ## 18830 ENSMUSG00000060216    Arrb2 ENSMUST00000102564
    ## 17765 ENSMUSG00000053819   Camk2d ENSMUST00000199300
    ## 4751  ENSMUSG00000022377    Asap1 ENSMUST00000177083
    ## 6137  ENSMUSG00000024969    Mark2 ENSMUST00000032557
    ## 4258  ENSMUSG00000021693    Kif2a ENSMUST00000022204
    ##              matchGeneID Firestar Matador3D CORSAIR SPADE THUMP
    ## 16712 ENSMUSG00000046707       20    10.130    17.3 239.6     0
    ## 18830 ENSMUSG00000060216       19    10.954    10.4 231.9     0
    ## 17765 ENSMUSG00000053819       22    13.150     7.4 460.0     1
    ## 4751  ENSMUSG00000022377        4    17.170     7.4 325.6     0
    ## 6137  ENSMUSG00000024969       18    12.208     3.9 355.5     0
    ## 4258  ENSMUSG00000021693       11    10.200    12.4 311.7     0
    ##       SignalP TargetP AS.Spearman.cor AS.Spearman.p     AS.Tau
    ## 16712      -4      -7      0.02366905             0 0.07826264
    ## 18830      -4      -6     -0.61539539             1 0.06630516
    ## 17765      -4      -5      0.67210377             1 0.13356087
    ## 4751       -4      -5      0.53255370             1 0.08022470
    ## 6137       -4      -5     -0.56214002             1 0.07963226
    ## 4258       -4      -5     -0.55622276             1 0.14939285
    ##        AS.Range      AS.Slope AS.Fit.p   ASDyScore
    ## 16712 0.1080385 -0.0005891379        0 -0.01708681
    ## 18830 0.1329095 -0.0055027170        1 -0.44786721
    ## 17765 0.2176835  0.0081598606        1  0.50230408
    ## 4751  0.1222690  0.0052077026        1  0.43633102
    ## 6137  0.2261130 -0.0060386351        1 -0.47463515
    ## 4258  0.2106085 -0.0064918103        1 -0.48535124
    ##       Gene.Spearman.cor Gene.Spearman.p  Gene.Tau Gene.Range
    ## 16712       -0.75149245               1 0.3314112     21.865
    ## 18830       -0.04733811               0 0.6528485     51.735
    ## 17765       -0.63314718               1 0.5671612     13.920
    ## 4751         0.15976611               0 0.4263838     25.710
    ## 6137        -0.57989181               1 0.3055633     18.570
    ## 4258         0.27811138               0 0.4362119     18.850
    ##       Gene.Slope Gene.Fit.p  GeDyScore FAScore FASType
    ## 16712 -0.4224782          1 -0.7508970 0.99817    Func
    ## 18830 -0.1395608          0 -0.3066246 0.99751    Func
    ## 17765 -0.6466652          1 -0.8078289 0.99697    Func
    ## 4751   0.1356671          0  0.2869695 0.99666    Func
    ## 6137  -0.1765882          0 -0.5103406 0.99657    Func
    ## 4258   0.0788720          0  0.2988659 0.99646    Func

or visualize GMM

``` r
calcuGMM(MyObj, LOG = 2, silent =T, main="Mouse")
```

    ## fitting ...
    ##|                                                      |   0%  
    |============================                            |  50%                                                                  |========================================================| 100%

</br>

<div align="left">
<img width = '350' height ='300' src ="https://github.com/LuChenLab/FAScore/blob/master/figure/GMM.png"/>
</div>

<br>

</br>  
</br>

## SessionInfo

``` r
sessionInfo()
```

    ## R version 4.1.2 (2021-11-01)
    ## Platform: x86_64-pc-linux-gnu (64-bit)
    ## Running under: Ubuntu 18.04.6 LTS
    ## 
    ## Matrix products: default
    ## BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
    ## LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
    ## 
    ## locale:
    ##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
    ##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
    ##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
    ##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
    ##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
    ## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
    ## 
    ## attached base packages:
    ## [1] parallel  stats4    stats     graphics  grDevices utils    
    ## [7] datasets  methods   base     
    ## 
    ## other attached packages:
    ##  [1] ggplot2_3.5.1               SummarizedExperiment_1.21.3
    ##  [3] Biobase_2.51.0              GenomicRanges_1.43.4       
    ##  [5] GenomeInfoDb_1.27.11        IRanges_2.25.7             
    ##  [7] S4Vectors_0.29.15           BiocGenerics_0.37.1        
    ##  [9] MatrixGenerics_1.3.1        matrixStats_0.62.0         
    ## [11] data.table_1.14.0           mclust_5.4.9               
    ## [13] dplyr_1.1.4                 FAScore_0.3.0              
    ## 
    ## loaded via a namespace (and not attached):
    ##  [1] httr_1.4.2               bit64_4.0.5             
    ##  [3] assertthat_0.2.1         highr_0.9               
    ##  [5] BiocFileCache_1.99.3     blob_1.2.2              
    ##  [7] GenomeInfoDbData_1.2.4   Rsamtools_2.7.2         
    ##  [9] yaml_2.2.1               progress_1.2.2          
    ## [11] pillar_1.9.0             RSQLite_2.2.7           
    ## [13] lattice_0.20-45          glue_1.6.2              
    ## [15] digest_0.6.27            XVector_0.31.1          
    ## [17] randomForest_4.6-14      colorspace_2.0-2        
    ## [19] htmltools_0.5.1.1        Matrix_1.6-0            
    ## [21] XML_3.99-0.6             pkgconfig_2.0.3         
    ## [23] biomaRt_2.47.7           zlibbioc_1.37.0         
    ## [25] scales_1.3.0             BiocParallel_1.25.5     
    ## [27] tibble_3.2.1             KEGGREST_1.31.1         
    ## [29] farver_2.1.0             generics_0.1.0          
    ## [31] ellipsis_0.3.2           withr_2.5.0             
    ## [33] cachem_1.0.5             GenomicFeatures_1.43.8  
    ## [35] cli_3.6.4                magrittr_2.0.1          
    ## [37] crayon_1.4.1             memoise_2.0.0           
    ## [39] evaluate_0.14            fansi_0.5.0             
    ## [41] textshaping_0.3.5        tools_4.1.2             
    ## [43] prettyunits_1.1.1        hms_1.1.0               
    ## [45] BiocIO_1.1.2             lifecycle_1.0.3         
    ## [47] stringr_1.5.1            munsell_0.5.0           
    ## [49] DelayedArray_0.17.10     AnnotationDbi_1.53.1    
    ## [51] Biostrings_2.59.2        compiler_4.1.2          
    ## [53] systemfonts_1.0.2        rlang_1.1.2             
    ## [55] grid_4.1.2               RCurl_1.98-1.3          
    ## [57] rstudioapi_0.13          rjson_0.2.20            
    ## [59] rappdirs_0.3.3           labeling_0.4.2          
    ## [61] bitops_1.0-7             rmarkdown_2.9           
    ## [63] restfulr_0.0.13          gtable_0.3.0            
    ## [65] DBI_1.1.1                curl_4.3.2              
    ## [67] R6_2.5.0                 GenomicAlignments_1.27.2
    ## [69] knitr_1.33               rtracklayer_1.51.5      
    ## [71] fastmap_1.1.0            bit_4.0.4               
    ## [73] utf8_1.2.2               filelock_1.0.2          
    ## [75] ragg_1.1.3               stringi_1.7.3           
    ## [77] Rcpp_1.0.12              vctrs_0.6.4             
    ## [79] png_0.1-7                xfun_0.24               
    ## [81] dbplyr_2.1.1             tidyselect_1.2.0

</br>  
</br> </br>

## Contact

Please contact Lu Chen (<luchen@scu.edu.cn>) or Li Chen (<chenli5679@163.com>).

</br>

## Citation

If you use FAScore in your publication, please cite FAScore by
